Fails if input data type is not string

When not using "by_value", the schema expects the enum to be of string type.

If the value is of different type, it does not detect error (like "Value must be of type str"), but it fails.

Following test case:

```python
from marshmallow import Schema, fields
from marshmallow_enum import EnumField
from enum import Enum


class SexEnum(Enum):
    male = 0
    female = 1


class SexyUserSchema(Schema):
    username = fields.String(required=True)
    sex = EnumField(SexEnum,
                    required=True,
                    error="{input} must be  one of {choices}")


def test_enum_wrong_type():
    user_schema = SexyUserSchema()

    indata = dict(username="jan", sex=324)

    res = user_schema.load(indata)
    print(res)
    assert res.errors
```

fails:


```
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>

    def test_enum_wrong_type():
        user_schema = SexyUserSchema()

        indata = dict(username="jan", sex=324)

>       res = user_schema.load(indata)

tests/test_enum_attnostring.py:23:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.tox/py36/lib/python3.6/site-packages/marshmallow/schema.py:580: in load
    result, errors = self._do_load(data, many, partial=partial, postprocess=True)
.tox/py36/lib/python3.6/site-packages/marshmallow/schema.py:660: in _do_load
    index_errors=self.opts.index_errors,
.tox/py36/lib/python3.6/site-packages/marshmallow/marshalling.py:295: in deserialize
    index=(index if index_errors else None)
.tox/py36/lib/python3.6/site-packages/marshmallow/marshalling.py:68: in call_and_store
    value = getter_func(data)
.tox/py36/lib/python3.6/site-packages/marshmallow/marshalling.py:288: in <lambda>
    data
.tox/py36/lib/python3.6/site-packages/marshmallow/fields.py:265: in deserialize
    output = self._deserialize(value, attr, data)
.tox/py36/lib/python3.6/site-packages/marshmallow_enum/__init__.py:31: in _deserialize
    return self._deserialize_by_name(value, attr, data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <fields.EnumField(default=<marshmallow.missing>, attribute=None, validate=None, required=True, lo
ad_only=False, dump_o...idator_failed': 'Invalid value.', 'by_name': 'Invalid enum member {input}', 'by_
value': 'Invalid enum value {input}'})>
value = 324, attr = 'sex', data = {'sex': 324, 'username': 'jan'}

    def _deserialize_by_name(self, value, attr, data):
        try:
>           return getattr(self.enum, value)
E           TypeError: getattr(): attribute name must be string

.tox/py36/lib/python3.6/site-packages/marshmallow_enum/__init__.py:41: TypeError
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/javl/sandbox/playwith_marshmallow/.tox/py36/lib/python3.6/site-packages/marshmallow_enum/__init_
_.py(41)_deserialize_by_name()
-> return getattr(self.enum, value)
```

--- a/marshmallow_enum/__init__.py
+++ b/marshmallow_enum/__init__.py
@@ -45,7 +45,7 @@ class EnumField(Field):
 
     def _deserialize_by_name(self, value, attr, data):
         if not isinstance(value, str_types):
-            self.fail('must_be_string')
+            self.fail('must_be_string', input=value)
 
         try:
             return getattr(self.enum, value)

--- a/marshmallow_enum/__init__.py
+++ b/marshmallow_enum/__init__.py
@@ -1,11 +1,18 @@
 from marshmallow.fields import Field
 from marshmallow import ValidationError
 
+try:
+    # ugh Python 2
+    str_types = (str, unicode)
+except:
+    str_types = (str,)
+
 
 class EnumField(Field):
     default_error_messages = {
         'by_name': 'Invalid enum member {input}',
-        'by_value': 'Invalid enum value {input}'
+        'by_value': 'Invalid enum value {input}',
+        'must_be_string': 'Enum name must be string'
     }
 
     def __init__(self, enum, by_value=False, error='',  *args, **kwargs):
@@ -37,6 +44,9 @@ class EnumField(Field):
             self.fail('by_value', input=value)
 
     def _deserialize_by_name(self, value, attr, data):
+        if not isinstance(value, str_types):
+            self.fail('must_be_string')
+
         try:
             return getattr(self.enum, value)
         except AttributeError:

--- a/tests/test_enum_field.py
+++ b/tests/test_enum_field.py
@@ -4,6 +4,9 @@ from marshmallow.fields import List
 from enum import Enum
 from collections import namedtuple
 import pytest
+import sys
+
+PY2 = sys.version_info.major == 2
 
 
 class EnumTester(Enum):
@@ -155,3 +158,39 @@ class TestCustomErrorMessage(object):
             field._deserialize(4, None, {})
         expected = 'Invalid enum value 4'
         assert expected in str(excinfo.value)
+
+
+class TestRegressions(object):
+    @pytest.mark.parametrize('bad_value', [
+        object, object(), [], {}, 1, 3.4, False, ()
+    ])
+    def test_by_name_must_be_string(self, bad_value):
+        class SomeEnum(Enum):
+            red = 0
+            yellow = 1
+            green = 2
+
+        class SomeSchema(Schema):
+            colors = EnumField(
+                SomeEnum,
+                by_value=False,
+            )
+
+        with pytest.raises(ValidationError) as excinfo:
+            SomeSchema(strict=True).load({'colors': bad_value})
+
+
+        assert 'must be string' in  str(excinfo.value)
+
+    @pytest.mark.skipif(PY2, reason='py2 strings are bytes')
+    def test_by_name_cannot_be_bytes(self):
+        class SomeEnum(Enum):
+            a = 1
+
+        class SomeSchema(Schema):
+            f = EnumField(SomeEnum, by_value=False)
+
+        with pytest.raises(ValidationError) as excinfo:
+            SomeSchema(strict=True).load({'f': b'a'})
+
+        assert 'must be string' in str(excinfo.value)

